lang/cs/Org.Apache.REEF.Wake/Avro/ProtocolSerializer.cs

```diff
@@ -0,0 +1,175 @@
+    /// class reflects all of the message classes and builds Avro seriailziers and deserializers.
+    /// Avro messages are then serialized/deserilaized to and from byte[] arrays using the
+    /// Read/Write methods. A transport such as a RemoteObserver using a ByteCodec can then
+    /// be used to send and receive the serialized messages.
+    /// </summary>
+    public sealed class ProtocolSerializer
+    {
+        private static readonly Logger Logr = Logger.GetLogger(typeof(ProtocolSerializer));
+
+        // Delagates for message serializers and deserializers.
+        private delegate void Serialize(MemoryStream stream, object message);
+        private delegate void Deserialize(MemoryStream stream, object observer, long sequence);
+
+        // Message type to serialize/derserialize delagate.
+        private SortedDictionary<string, Serialize> serializeMap = new SortedDictionary<string, Serialize>();
+        private SortedDictionary<string, Deserialize> deserializeMap = new SortedDictionary<string, Deserialize>();
+
+        /// <summary>
+        /// Register all of the protocol messages using reflection.
+        /// </summary>
+        /// <param name="assembly">The Assembley object which contains the namespace of the message classes.</param>
+        /// <param name="messageNamespace">A string which contains the namespace the protocol messages.</param>
+        public void Initialize(Assembly assembly, string messageNamespace)
+        {
+            MethodInfo registerInfo = typeof(ProtocolSerializer).GetMethod("Register", BindingFlags.Instance | BindingFlags.NonPublic);
+            MethodInfo genericInfo;
+
+            Logr.Log(Level.Info, string.Format("Retrieving types for assembly: {0}", assembly.FullName));
+            Type[] types = assembly.GetTypes();
+            foreach (Type type in types)
+            {
+                string name = type.FullName;
+                if (name.Contains(messageNamespace))
+                {
+                    genericInfo = registerInfo.MakeGenericMethod(new[] { type });
+                    genericInfo.Invoke(this, null);
+                }
+            }
+
+            // Register the header message.
+            genericInfo = registerInfo.MakeGenericMethod(new[] { typeof(Header) });
+            genericInfo.Invoke(this, null);
+        }
+
+        /// <summary>
+        /// Generate and store the metadata necessary to serialze and deserialize a specific message type.
+        /// </summary>
+        /// <typeparam name="TMessage">The class type of the message being registered.</typeparam>
+        internal void Register<TMessage>()
+        {
+            Logr.Log(Level.Info, string.Format("Registering message type: {0} {1}", typeof(TMessage).FullName, typeof(TMessage).Name));
+
+            Serialize serialize = (MemoryStream stream, object message) =>
+            {
+                IAvroSerializer<TMessage> messageWriter = AvroSerializer.Create<TMessage>();
```

Comment:
I was aware of the memory allocation and thinking about the best way to solve it. Docs show the lambda reference will for the serializer to continue to exist.
Done